### PR TITLE
server: register multiple api handlers

### PR DESCRIPTION
**Pull request by nolouch**
Signed-off-by: nolouch <nolouch@gmail.com>

### What problem does this PR solve? 
Supports to register some resource handler to the server, such as `keyvisual`.

### What is changed and how it works?
There are some new introduce about resource groups like [k8s defined](https://kubernetes.io/docs/concepts/overview/kubernetes-api/#api-groups).

### Check List 

Tests 

 - Unit test
 - Integration test

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -101,8 +103,54 @@ type Server struct {
 	logProps *log.ZapProperties
 }
 
+// HandlerBuilder build a server HTTP handler.
```
Comment (disksing) [inline server/server.go]: builds

File: server/server.go [hunk 2]
```diff
@@ -101,8 +103,54 @@ type Server struct {
...
+	registerMap := make(map[string]struct{})
+	for _, b := range apiBuilders {
+		h, info := b(svr)
+		var p string
+		if info.IsCore {
+			p = CorePath
+		} else {
+			p = path.Join(ExtensionsPath, info.Group, info.Version)
+		}
+		if _, ok := registerMap[p]; ok {
```
Comment (disksing) [inline server/server.go]: map always empty.

**Comment by nolouch** [comment]
PTAL  @shafreeck @lhy1024

File: server/server.go [hunk 3]
```diff
@@ -101,8 +103,53 @@ type Server struct {
 	logProps *log.ZapProperties
 }
 
+// HandlerBuilder builds a server HTTP handler.
+type HandlerBuilder func(*Server) (http.Handler, APIGroupInfo)
+
+// APIGroupInfo used to regster the api service.
+type APIGroupInfo struct {
```
Comment (shafreeck) [inline server/server.go]: How about just call it `APIGroup`?

File: server/server.go [hunk 4]
```diff
@@ -101,8 +103,53 @@ type Server struct {
...
+	IsCore  bool
+	Group   string
+	Version string
+}
+
+const (
+	//CorePath  the core group, is at REST path `/pd/api/v1`.
+	CorePath = "/pd/api/v1"
+	// ExtensionsPath the named groups are REST at `/pd/apis/{GROUP_NAME}/{Version}`.
+	ExtensionsPath = "/pd/apis"
```
Comment (shafreeck) [inline server/server.go]: How to understand `api` and `apis`, maybe use `/pd/ext/`, `/pd/extension/` or `/pd/plugin/` ?

File: server/server.go (hunk 3 above)
Comment (shafreeck) [reply server/server.go]: For Example 
```
type APIGroup struct {
    Name string
    Version string
    IsCore bool
}
```

File: server/server.go [hunk 5]
```diff
@@ -101,8 +103,53 @@ type Server struct {
...
+)
+
+func combineBuilderServerHTTPService(svr *Server, apiBuilders ...HandlerBuilder) (http.Handler, error) {
+	engine := negroni.New()
+	recovery := negroni.NewRecovery()
+	engine.Use(recovery)
+	router := mux.NewRouter()
+	registerMap := make(map[string]struct{})
+	for _, b := range apiBuilders {
```
Comment (lhy1024) [inline server/server.go]: would you like to replace `h``b``p` with meaningful name?

File: server/server.go [hunk 6]
```diff
@@ -101,8 +103,53 @@ type Server struct {
 	logProps *log.ZapProperties
 }
 
+// HandlerBuilder builds a server HTTP handler.
+type HandlerBuilder func(*Server) (http.Handler, APIGroupInfo)
+
+// APIGroupInfo used to regster the api service.
```
Comment (lhy1024) [inline server/server.go]:
```suggestion
// APIGroupInfo used to register the api service.
```

File: server/server.go [hunk 7]
```diff
@@ -101,8 +103,53 @@ type Server struct {
...
+	for _, b := range apiBuilders {
+		h, info := b(svr)
+		var p string
+		if info.IsCore {
+			p = CorePath
+		} else {
+			p = path.Join(ExtensionsPath, info.Group, info.Version)
+		}
+		if _, ok := registerMap[p]; ok {
+			return nil, errors.Errorf("service with path [%s] alredy registered", p)
```
Comment (lhy1024) [inline server/server.go]:
```suggestion
			return nil, errors.Errorf("service with path [%s] already registered", p)
```

File: server/server.go [hunk 8]
```diff
@@ -101,8 +103,53 @@ type Server struct {
...
+		registerMap[p] = struct{}{}
+		router.PathPrefix(p).Handler(h)
+	}
+	engine.UseHandler(router)
+	return engine, nil
+}
+
 // CreateServer creates the UNINITIALIZED pd server with given configuration.
-func CreateServer(cfg *config.Config, apiRegister func(*Server) http.Handler) (*Server, error) {
+func CreateServer(cfg *config.Config, apiBulders ...HandlerBuilder) (*Server, error) {
```
Comment (lhy1024) [inline server/server.go]:
```suggestion
func CreateServer(cfg *config.Config, apiBuilders ...HandlerBuilder) (*Server, error) {
```

File: server/server.go (hunk 4 above)
Comment (nolouch) [reply server/server.go]: @shafreeck  I prefer all API follower this style `/pd/apis/{GROUP_NAME}/{Version}`, but we have a history of burdens. the group may not only the extension but also some core components like each scheduler. so do you have a suggestion?

File: server/server.go [hunk 9]
```diff
@@ -101,8 +103,53 @@ type Server struct {
...
+)
+
+func combineBuilderServerHTTPService(svr *Server, apiBuilders ...HandlerBuilder) (http.Handler, error) {
+	engine := negroni.New()
+	recovery := negroni.NewRecovery()
+	engine.Use(recovery)
+	router := mux.NewRouter()
+	registerMap := make(map[string]struct{})
+	for _, build := range apiBuilders {
+		hander, info := build(svr)
```
Comment (lhy1024) [inline server/server.go]:
```suggestion
		handler, info := build(svr)
```

**Pull request merged**